helix-term/src/commands.rs

```diff
@@ -980,7 +1042,25 @@ fn split_selection_on_newline(cx: &mut Context) {
 fn search_impl(doc: &mut Document, view: &mut View, contents: &str, regex: &Regex, extend: bool) {
     let text = doc.text();
     let selection = doc.selection(view.id);
-    let start = text.char_to_byte(selection.cursor());
+    let start = {
+        let range = selection.primary();
+
+        // This is a little bit weird.  Due to 1-width cursor semantics, we
+        // would typically want the search to always begin at the visual left-side
+        // of the head.  However, when there's already a selection from e.g. a
+        // previous search result, we don't want to include any of that selection
+        // in the subsequent search.  The code below makes a compromise between the
+        // two behaviors that hopefully behaves the way most people expect most of
+        // the time.
+        if range.anchor <= range.head {
```

Comment:
Let's make a function for this `range.is_reverse`/`is_inverse`?

Reply:
Generally I agree with you, and I'll break that out where those are the actual semantics.  In this case that's not what it's checking (note that it's `<=` not `<`).

I also just remembered that this is a case that's actually taking advantage of zero-width cursors in a subtle way, and won't work if we do min-width globally.  In master if you have text like this:

```
aaaaaaaa
```

And you do a single-character search for `a`, then pressing `n` won't progress, it just stays stuck on the first `a`.  And if you do a double-character search for `aa` and press `n` it actually progresses through all overlapping results.  Neither of those behaviors are what I expected, at least.  And at the very least the first behavior seems broken.

This code uses whether the range is effectively zero-width or not (sort of... that's what the `<=` check actually is) to determined where to start the search, so that things behave as expected.  Searching for just `a` will still start on the current character, but pressing `n` will cycle through subsequent characters.  And searching for `aa` will progress through non-overlapping matches.

I think this kind of thing is where preserving zero-width cursors behind the scenes can actually make behaviors match people's expectations *better*, even if the user is never aware of what's going on behind the scenes and still sees everything as block cursors.  I'm hoping to experiment with more stuff like that after this PR, but I forgot that I already took advantage of it here to fix some behavior that wasn't just unexpected, but actually seems broken(?) otherwise.  Specially, not progressing through single-character searches.

Thoughts?

Reply:
After discussing on Matrix, we've decided to follow vim's behavior and just always start searches to the right of the block cursor.